File: examples/readmeexample.py
"""
Full version of the GitHub README.
"""

from tgintegration import BotIntegrationClient

# This example uses the configuration of `config.ini` (see examples/README)
client = BotIntegrationClient(
    bot_under_test='@BotListBot',
    session_name='my_account',  # Arbitrary file path to the Pyrogram session file
    max_wait_response=8,  # Maximum timeout for bot responses
    min_wait_consecutive=2,  # Minimum time to wait for consecutive messages
    raise_no_response=True  # Raise `InvalidResponseError` when no response received
)

client.start()
client.clear_chat()  # Let's start with a blank screen

### Now let's send the /start command to the bot_under_test and "await" exactly three messages:
response = client.send_command_await("start", num_expected=3)

assert response.num_messages == 3
assert response.messages[0].sticker  # First message is a sticker

### Let's examine these buttons in the response...
inline_keyboard = response.inline_keyboards[0]
# Three buttons in the first row
assert len(inline_keyboard.rows[0]) == 3

### We can also query and press the inline keyboard buttons:
# Click the first button matching the pattern
examples = response.inline_keyboards[0].press_button_await(pattern=r'.*Examples')

assert "Examples for contributing to the BotList" in examples.full_text
### As the bot edits the message, `press_inline_button` automatically listens for `MessageEdited`
### events and picks up on the edit, returning it as `Response`.

### So what happens when we send an invalid query or the bot fails to respond?
from tgintegration import InvalidResponseError
try:
    # The following instruction will raise an `InvalidResponseError` after
    # `client.max_wait_response` seconds. This is because we passed `raise_no_response = True`
    # in the client initialization.
    client.send_command_await("ayylmao", raise_=True)
except InvalidResponseError:
    print("Raised.")  # Ok

### The `BotIntegrationClient` is based off a regular Pyrogram Client, meaning that, in addition to
###  the `send_*_await` methods, all normal Pyro methods still work:
client.send_message(client.bot_under_test, "Hello from Pyrogram")  # Not waiting for response

# `send_*_await` methods automatically use the `bot_under_test` as peer:
res = client.send_message_await("Hello from TgIntegration", max_wait=2, raise_=False)
# If `raise_` is explicitly set to False, no exception is raised:
assert res.empty
# Note that when no response is expected and no validation thereof is necessary, ...
client.send_photo_await("media/photo.jpg", max_wait=0, raise_=False)
client.send_voice_await("media/voice.ogg", max_wait=0, raise_=False)
# ... it makes more sense to use the "unawaitable" methods:
client.send_photo(client.bot_under_test, "media/photo.jpg")
client.send_voice(client.bot_under_test, "media/voice.ogg")

### Custom awaitable Actions
from tgintegration import AwaitableAction, Response
from pyrogram import Filters

peer = '@BotListChat'

action = AwaitableAction(
    func=client.send_message,
    kwargs=dict(
        chat_id=peer,
        text="**Hello World**",
        parse_mode='markdown'
    ),
    # Wait for messages only by the peer we're interacting with
    filters=Filters.user(peer) & Filters.incoming,
    # Time out and raise after 15 seconds
    max_wait=15
)

response = client.act_await_response(action)  # type: Response

File: examples/readme-example.ts
import { ChatController, InvalidResponseError } from "@tgintegration/core";
import { createClient } from "./utils.js";

/**
 * Full version of the GitHub README example.
 */
async function main() {
  const client = await createClient();

  // We are going to run tests on https://t.me/BotListBot (or any other bot)
  // Note: BotListBot might be down, replace with your target bot.
  const peer = "@BotListBot";

  const controller = new ChatController(client, peer, {
    globalActionDelay: 2500, // 2.5s delay to follow along
  });

  await controller.initialize();

  console.log("Clearing chat to start with a blank screen...");
  await controller.clearChat();

  console.log("Sending /start and waiting for exactly 3 messages...");
  const response = await controller.collect(
    {
      minMessages: 3,
      maxMessages: 3,
      maxWait: 8000,
    },
    async () => {
      await controller.sendCommand("start");
    },
  );

  if (response.count !== 3) {
    throw new Error(`Expected 3 messages, got ${response.count}`);
  }
  console.log("Three messages received, bundled as a `Response`.");

  // Check for sticker
  if (response.messages[0].media?.type === "sticker") {
    console.log("First message is a sticker.");
  }

  console.log("Let's examine the buttons in the response...");
  const inlineKeyboards = response.inlineKeyboards;
  if (inlineKeyboards.length > 0) {
    const kb = inlineKeyboards[0];
    // In legacy code, rows[0] had 3 buttons.
    if (kb.buttons[0].length === 3) {
      console.log("Yep, there are three buttons in the first row.");
    }

    // Click button by pattern
    console.log("Clicking button matching 'Examples'...");
    try {
      const clickResult = await kb.click(/.*Examples/);

      if (clickResult.answered) {
        console.log("Button click was acknowledged by the bot!");
      }
    } catch (_e) {
      console.log(
        "Could not find or click button (bot might be down or changed).",
      );
    }
  }

  console.log(
    "So what happens when we send an invalid query or the peer fails to respond?",
  );

  try {
    console.log("Expecting unhandled command to raise InvalidResponseError...");
    await controller.collect(
      {
        maxWait: 8000,
        throwOnTimeout: true,
      },
      async () => {
        await controller.sendCommand("ayylmao");
      },
    );
  } catch (e) {
    if (e instanceof InvalidResponseError) {
      console.log("Ok, raised as expected.");
    } else {
      throw e;
    }
  }

  // If `throwOnTimeout` is false (default), no exception is raised
  const noReplyResponse = await controller.collect(
    {
      maxWait: 3000,
      throwOnTimeout: false,
    },
    async () => {
      console.log("Sending a message but expecting no reply...");
      // Send directly via client if needed, or use controller
      await controller.client.sendText(controller.peer, "Henlo Fren");
    },
  );

  if (noReplyResponse.count === 0) {
    console.log("Success! No response received as expected.");
  }
}

await main();
